File: src/middleware/middleware.go
package middleware

import (
	"fmt"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
	"kinozal-bot/config"
	"kinozal-bot/logger"
)

type AccessMiddleware struct {
	Bot *tgbotapi.BotAPI
	Cfg *config.Config
}

func (am *AccessMiddleware) CheckAccess(userID int, chatID int64) bool {
	isAllowed := am.isAllowedUser(userID)

	if !isAllowed {
		logger.Warn("Unauthorized access attempt", map[string]interface{}{
			"user_id": userID,
			"chat_id": chatID,
		})

		msg := tgbotapi.NewMessage(chatID, "У вас нет доступа к этому боту. Пожалуйста, обратитесь к администратору.")
		_, err := am.Bot.Send(msg)
		if err != nil {
			logger.Error("Failed to send access denial message", map[string]interface{}{
				"error": err.Error(),
			})
		}

		if am.Cfg.Bot.AdminID != 0 {
			adminMsg := fmt.Sprintf("⚠️ Попытка несанкционированного доступа:\nUser ID: %d\nChat ID: %d", userID, chatID)
			adminMessage := tgbotapi.NewMessage(int64(am.Cfg.Bot.AdminID), adminMsg)
			_, err := am.Bot.Send(adminMessage)
			if err != nil {
				logger.Warn("Failed to notify admin about unauthorized access", map[string]interface{}{
					"error": err.Error(),
				})
			}
		}
	}

	return isAllowed
}

func (am *AccessMiddleware) isAllowedUser(userID int) bool {
	if userID == am.Cfg.Bot.AdminID {
		return true
	}

	for _, allowedID := range am.Cfg.Bot.AllowedUsers {
		if allowedID == userID {
			return true
		}
	}

	return false
}

File: JS_version/middleware.js
class AccessMiddleware {
    constructor(config, bot, logger) {
        this.config = config;
        this.bot = bot;
        this.logger = logger;
    }

    /**
     * Check if user is allowed to use the bot
     * @param {number} userId - Telegram user ID
     * @param {number} chatId - Telegram chat ID
     * @returns {boolean} - Whether user is allowed
     */
    async checkAccess(userId, chatId) {
        const isAllowed = this.config.isAllowedUser(userId);
        
        if (!isAllowed) {
            this.logger.warn('Unauthorized access attempt', {
                userId,
                chatId
            });
            
            await this.bot.sendMessage(
                chatId,
                'У вас нет доступа к этому боту. Обратитесь к администратору.'
            );
            
            // Notify admin about unauthorized access attempt
            if (this.config.bot.adminId) {
                await this.bot.sendMessage(
                    this.config.bot.adminId,
                    `⚠️ Попытка несанкционированного доступа:\nUser ID: ${userId}\nChat ID: ${chatId}`
                );
            }
        }
        
        return isAllowed;
    }
}
module.exports = AccessMiddleware;
